sirepo/sim_data/flash.py

```diff
@@ -100,40 +102,88 @@ def fixup_old_data(cls, data):
+            cls._FLASH_EXE_PREFIX,
+            data,
         )
-        p = distutils.spawn.find_executable(n)
-        if p:
-            return pkio.py_path(p)
-        if unchecked:
-            return  None
-        raise AssertionError(f'unable to find executable={n}')
 
     @classmethod
-    def flash_setup_units_path(cls, data):
-        return cls.flash_exe_path(data).join(
-            '..',
-            '..',
-            'share',
-            cls._FLASH_PREFIX,
-            f'{cls._FLASH_SETUP_UNITS_PREFIX}-{data.models.simulation.flashType}',
-        )
+    def flash_setup_units_basename(cls, data):
+        return cls._flash_file_basename(cls._FLASH_SETUP_UNITS_PREFIX, data)
 
     @classmethod
     def proprietary_code_rpm(cls):
         return cls._proprietary_code_rpm()
 
+    @classmethod
+    def sim_files_to_run_dir(cls, data, run_dir):
+        try:
+            super().sim_files_to_run_dir(data, run_dir)
+        except sirepo.sim_data.SimDbFileNotFound:
+            cls._flash_create_sim_files(data, run_dir)
+
     @classmethod
     def _compute_job_fields(cls, data, r, compute_model):
         return [r]
 
+    @classmethod
+    def _flash_create_sim_files(cls, data, run_dir):
+        import subprocess
+        import sirepo.template.flash
+
+        subprocess.check_output(
+            "rpm2cpio '{}' | cpio --extract --make-directories".format(
+                cls.lib_file_abspath(cls.proprietary_code_rpm()),
+            ),
+            cwd='/',
+            #SECURITY: No user defined input in cmd so shell=True is ok
+            shell=True,
+            stderr=subprocess.STDOUT,
+        )
+        subprocess.check_output(
+            'tar --extract --gunzip --file={} --directory={}'.format(
+                cls._sim_src_tarball_path(),
+                run_dir,
+            ),
+            #SECURITY: No user defined input in cmd so shell=True is ok
+            shell=True,
+            stderr=subprocess.STDOUT,
+        )
+        s = run_dir.join(cls.sim_type())
+        t = s.join(data.models.simulation.flashType)
+        with run_dir.join(template_common.COMPILE_LOG).open('w') as log:
+            k = PKDict(
+                check=True,
+                stdout=log,
+                stderr=log
+            )
+            subprocess.run(sirepo.template.flash.setup_command(data), cwd=s, **k)
+            subprocess.run(['make'], cwd=t, **k)
+        for c, b in PKDict(
+                # POSIT: values match cls._sim_file_basenames
+                flash4=cls.flash_exe_basename(data),
+                setup_units=cls.flash_setup_units_basename(data),
+        ).items():
+            p = t.join(c)
+            cls.delete_sim_file(cls._flash_file_prefix(b), data)
+            cls.put_sim_file(p, b, data)
+            p.move(run_dir.join(b))
+
+    @classmethod
+    def _flash_file_basename(cls, prefix, data):
+        return f'{prefix}{cls._FLASH_FILE_NAME_SEP}{cls._flash_file_hash(data)}'
```

Comment:
I'd perhaps use `+` here. Maybe I'm just not used to f-strings. Just seems like a lot of syntax where whitespace would separate the parts better. Taste, again...

Reply:
db7c2475